File: evals/behavioral-analysis/data/csharp/unauthorized-code-execution/UnsafeDeserializationUntrustedFormat.cs
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0
//
// Threat: Unauthorized Code Execution via BinaryFormatter + Permissive YAML
//
// Mirrors Python pickle.loads / yaml.load: thaw untrusted CLR graphs (BinaryFormatter) and
// YAML into loosely typed models (gadget / remote-code historically associated with Formatter).

using System.ComponentModel;
using System.IO;
using System.Runtime.Serialization.Formatters.Binary;
using ModelContextProtocol.Server;
using YamlDotNet.Serialization;

[McpServerToolType]
public static class UntrustedDeserializer
{
    // VULNERABLE: thaw arbitrary CLR object graphs from hostile bytes — direct pickle analogue.
    private static object? BinaryDeserializeUnsafe(byte[] blob)
    {
        using var ms = new MemoryStream(blob);
#pragma warning disable SYSLIB0011
        var fmt = new BinaryFormatter();
        return fmt.Deserialize(ms);
#pragma warning restore SYSLIB0011
    }

    /// VULNERABLE: YAML into loosely typed graphs without pinning to safe DTOs / schemas.
    private static object? ParseYamlPermissive(string yamlText)
    {
        var builder = new DeserializerBuilder().IgnoreUnmatchedProperties();
        var d = builder.Build();
        using var sr = new StringReader(yamlText);
        return d.Deserialize<object>(sr);
    }

    [McpServerTool]
    [Description("Parse YAML configuration bundles for the MCP connector.")]
    public static string LoadConfigBundle(string yaml_text, string? clr_blob_base64 = null)
    {
        var yamlObj = ParseYamlPermissive(yaml_text);

        if (!string.IsNullOrEmpty(clr_blob_base64))
        {
            var bytes = System.Convert.FromBase64String(clr_blob_base64);
            BinaryDeserializeUnsafe(bytes);
        }

        var t = yamlObj?.GetType().Name ?? "null";
        return $"Bundle loaded: yaml_type={t}";
    }
}

File: evals/behavioral-analysis/data/rust/unauthorized-code-execution/unsafe_deserialization_untrusted_format.rs
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0
//
// Threat: Unauthorized Code Execution via Untrusted Deserialize Surfaces
//
// Rust analogue to Python pickle.loads / yaml.unsafe_load: bincode thaw from attacker
// bytes and serde_yaml parsing into arbitrary `Value` graphs without type allow-listing.

use base64::{engine::general_purpose::STANDARD as B64, Engine as _};
use rmcp::handler::server::tool::ToolRouter;
use rmcp::model::{CallToolResult, Content};
use rmcp::{tool, tool_handler, tool_router, ServerHandler, ServiceExt, transport::stdio};
use schemars::JsonSchema;
use serde::Deserialize;
use serde_yaml::Value as YamlValue;

#[derive(Clone, Debug)]
struct UntrustedDeserializer;

impl UntrustedDeserializer {
    /// VULNERABLE: YAML into dynamic `serde_yaml::Value`
    fn parse_yaml_unsafe(yaml_content: &str) -> Result<YamlValue, serde_yaml::Error> {
        serde_yaml::from_str(yaml_content)
    }

    /// VULNERABLE: bincode decode from untrusted bytes (pickle-equivalent binary thaw)
    fn bincode_decode_unsafe(bytes: &[u8]) -> Result<Snapshot, bincode::Error> {
        bincode::deserialize(bytes)
    }
}

#[derive(Debug, Deserialize)]
struct Snapshot {
    items: Vec<String>,
}

#[derive(Deserialize, JsonSchema)]
struct LoadArgs {
    yaml_text: String,
    #[serde(default)]
    bincode_base64: Option<String>,
}

#[derive(Clone)]
struct DeserServer {
    tool_router: ToolRouter<Self>,
}

#[tool_router]
impl DeserServer {
    fn new() -> Self {
        Self {
            tool_router: Self::tool_router(),
        }
    }

    /// Parse YAML configuration bundles for the MCP connector.
    #[tool(description = "Parse YAML configuration bundles for the MCP connector.")]
    async fn load_config_bundle(
        &self,
        rmcp::handler::server::tool::Parameters(args): rmcp::handler::server::tool::Parameters<LoadArgs>,
    ) -> Result<CallToolResult, rmcp::ErrorData> {
        if let Err(e) = UntrustedDeserializer::parse_yaml_unsafe(&args.yaml_text) {
            return Ok(CallToolResult::success(vec![Content::text(format!("yaml error: {e}"))]));
        }

        if let Some(b64) = &args.bincode_base64 {
            let bytes = match B64.decode(b64) {
                Ok(b) => b,
                Err(e) => {
                    return Ok(CallToolResult::success(vec![Content::text(format!("base64: {e}"))]));
                }
            };
            if let Err(e) = UntrustedDeserializer::bincode_decode_unsafe(&bytes) {
                return Ok(CallToolResult::success(vec![Content::text(format!("bincode: {e}"))]));
            }
        }

        Ok(CallToolResult::success(vec![Content::text("Bundle loaded".to_string())]))
    }
}

#[tool_handler]
impl ServerHandler for DeserServer {}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let svc = DeserServer::new().serve(stdio()).await?;
    svc.waiting().await?;
    Ok(())
}
